test/src/controllers/plugin.test.js

```diff
@@ -0,0 +1,213 @@
+import { expect } from "chai";
+import Controller from "../../../src/controllers/index";
+import PostgresGateway from "../../../src/gateways/postgres/index";
+import sinon from "sinon";
+
+describe("Controller Tests", function () {
+  const controller = new Controller();
+  const error = new Error("Fake Error");
+
+  describe("Write Action", function () {
+    const product = {
+      product_id: 1,
+      style_id: 1,
+      size: 13.5,
+      quantity: 10,
+    };
+
+    beforeEach(function () {
+      sinon.restore();
+    });
+
+    afterEach(function () {
+      sinon.restore();
+    });
+
+    it("should write add a new product to the database", async function () {
+      sinon
+        .stub(PostgresGateway.prototype, "write")
+        .callsFake(async function () {
+          const result = {
+            rows: [
+              {
+                product_id: "1",
+              },
+            ],
+          };
+
+          return result;
+        });
+
+      const response = await controller.write(product);
+
+      expect(Number(response.rows[0].product_id)).to.equal(1);
+    });
+
+    it("should throw an error if the method fails and log it with Winston", async function () {
+      const logger = sinon.spy(controller.gateway.logger, "log");
+
+      try {
+        sinon.stub(PostgresGateway.prototype, "write").throws(error);
+
+        await controller.write(product);
+      } catch (error) {
+        expect(error.message).to.equal("Fake Error");
+        sinon.assert.calledOnce(logger);
```

Comment:
**Annotation:** Since the error stub expect clause will always pass, it is wise to add a spy to ensure the logger has been called. (The logger will only be called when an error occurs).